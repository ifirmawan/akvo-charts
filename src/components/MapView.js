// src/MapView.js
import React, {
  useEffect,
  useRef,
  forwardRef,
  useImperativeHandle,
  useState,
  useCallback
} from 'react';
import L from 'leaflet';
import * as topojson from 'topojson-client';
import 'leaflet/dist/leaflet.css';
import markerIcon from 'leaflet/dist/images/marker-icon.png';
import markerShadow from 'leaflet/dist/images/marker-shadow.png';

// Define the default icon for markers
const defaultIcon = L.icon({
  iconUrl: typeof markerIcon === 'object' ? markerIcon?.src : markerIcon,
  shadowUrl:
    typeof markerShadow === 'object' ? markerShadow?.src : markerShadow,
  iconSize: [25, 41],
  iconAnchor: [12, 41],
  popupAnchor: [1, -34],
  shadowSize: [41, 41]
});

const getObjectFromString = (path) => {
  const obj = path.split('.').reduce((obj, key) => obj && obj[key], window);
  if (typeof obj === 'undefined' || typeof obj === 'string') {
    return null;
  }
  return obj;
};

const MapView = forwardRef(({ tile, layer, config, data = [] }, ref) => {
  const [geoData, setGeoData] = useState(null);

  const mapContainerRef = useRef(null);
  const mapInstanceRef = useRef(null);

  const loadGeoDataFromURL = useCallback(async () => {
    if (layer?.url && !geoData) {
      try {
        const res = await fetch(layer.url);
        const apiData = await res.json();
        if (apiData) {
          setGeoData(apiData);
        }
      } catch (err) {
        console.error('loadGeoDataFromURL', err);
      }
    }
  }, [layer.url, geoData]);

  useEffect(() => {
    loadGeoDataFromURL();
  }, [loadGeoDataFromURL]);

  useEffect(() => {
    if (mapInstanceRef.current === null && mapContainerRef.current) {
      // Initialize the map only if it hasn't been initialized yet
      const map = L.map(mapContainerRef.current, {
        center: config?.center || [0, 0],
        zoom: config?.zoom || 2
      });

      // Save the map instance to ref
      mapInstanceRef.current = map;

      // Add a tile layer to the map
      if (tile?.url) {
        const { url: tileURL, ...tileProps } = tile;
        L.tileLayer(tileURL, { ...tileProps }).addTo(map);
      }

      // Add a marker to the map
      data
        ?.filter((d) => d?.point && d?.label)
        ?.forEach((d) =>
          L.marker(d?.point, { icon: defaultIcon })
            .bindPopup(d?.label)
            .addTo(map)
        );
      // Create the TopoJSON layer
      const TopoJSON = L.GeoJSON.extend({
        addData: (d) => {
          if (d?.type === 'Topology') {
            for (let kd in d.objects) {
              if (d.objects.hasOwnProperty(kd)) {
                const geojson = topojson.feature(d, d.objects[kd]);
                L.geoJSON(geojson, { style: () => layer?.style || {} }).addTo(
                  map
                );
              }
            }
          }
          // Make sure `d` is an object that has a type but is not equal to "Topology"
          if (d?.type && d?.type !== 'Topology') {
            L.geoJSON(d, { style: () => layer?.style || {} }).addTo(map);
          }
        }
      });

      L.topoJson = function (d, options) {
        return new TopoJSON(d, options);
      };

      // Create an empty GeoJSON layer with a style and a popup on click
      const geojsonLayer = L.topoJson(null).addTo(map);

      try {
        // Set GeoJSON when source is string as window variable name
        if (
          typeof layer?.source === 'string' &&
          layer?.source?.includes('window')
        ) {
          const topoData = getObjectFromString(layer.source);
          if (topoData) {
            geojsonLayer.addData(topoData);
          }
        }
        // Set GeoJSON directly if the source is an object
        if (typeof layer?.source === 'object') {
          geojsonLayer.addData(layer.source);
        }
      } catch (err) {
        console.error('geojsonLayer', err);
      }

      if (geoData) {
        geojsonLayer.addData(geoData);
      }
    }

    // Cleanup function to remove map instance on unmount
    return () => {
      if (mapInstanceRef.current) {
        mapInstanceRef.current.remove();
        mapInstanceRef.current = null;
      }
    };
  }, [
    config?.center,
    config?.zoom,
    data,
    layer.source,
    layer?.style,
    layer.url,
    tile,
    geoData
  ]);

  useImperativeHandle(ref, () => ({
    zoomIn: () => {
      if (mapInstanceRef.current) {
        mapInstanceRef.current.zoomIn();
      }
    },
    zoomOut: () => {
      if (mapInstanceRef.current) {
        mapInstanceRef.current.zoomOut();
      }
    },
    getCenter: () => {
      if (mapInstanceRef.current) {
        return mapInstanceRef.current.getCenter();
      }
      return null;
    }
  }));

  return (
    <div
      ref={mapContainerRef}
      style={{
        height: config?.height || '100vh',
        width: config?.width || '100%'
      }}
      data-testid="map-view"
    />
  );
});

export default MapView;
