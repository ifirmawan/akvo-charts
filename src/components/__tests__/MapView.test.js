import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import MapView from '../MapView';

describe('MapView chart', () => {
  test('renders MapView correctly', async () => {
    const props = {
      tile: {
        url: 'https://tile.openstreetmap.org/{z}/{x}/{y}.png',
        maxZoom: 19,
        attribution: '© OpenStreetMap'
      },
      layer: {
        source: 'window.topoData'
      },
      data: [
        {
          point: [39.61, -105.02],
          label: 'This is Littleton, CO.'
        },
        {
          point: [39.73, -104.8],
          label: 'This is Aurora, CO.'
        }
      ],
      config: {
        center: [39.73, -104.99],
        zoom: 10,
        height: '100vh',
        width: '100%'
      }
    };
    let instance = null;
    render(
      <MapView
        {...props}
        ref={(el) => {
          instance = el;
        }}
      />
    );

    await waitFor(() => {
      const chartContainer = screen.getByTestId('map-view');
      expect(chartContainer).toBeInTheDocument();
      expect(instance.getCenter()).toEqual({
        lat: 39.73,
        lng: -104.99
      });
    });
  });

  test('matches MapView snapshot', async () => {
    const props = {
      tile: {
        url: 'https://tile.openstreetmap.org/{z}/{x}/{y}.png',
        maxZoom: 19,
        attribution: '© OpenStreetMap'
      },
      layer: {
        source: 'window.topoData'
      },
      data: [
        {
          point: [39.61, -105.02],
          label: 'This is Littleton, CO.'
        },
        {
          point: [39.73, -104.8],
          label: 'This is Aurora, CO.'
        }
      ],
      config: {
        center: [39.73, -104.99],
        zoom: 10,
        height: '100vh',
        width: '100%'
      }
    };

    let instance = null;
    const { container } = render(
      <MapView
        {...props}
        ref={(el) => {
          instance = el;
        }}
      />
    );
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
