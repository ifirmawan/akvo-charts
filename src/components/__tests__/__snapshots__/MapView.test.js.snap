// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapView chart matches MapView snapshot 1`] = `
<div>
  <div
    class="leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom"
    data-testid="map-view"
    style="height: 100vh; width: 100%; position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 19; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              src="https://tile.openstreetmap.org/10/213/388.png"
              style="width: 256px; height: 256px; left: -93px; top: -170px;"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      >
        <img
          alt=""
          class="leaflet-marker-shadow leaflet-zoom-hide"
          src="marker-shadow.png"
          style="margin-left: -12px; margin-top: -41px; width: 41px; height: 41px; left: -22px; top: 114px;"
        />
        <img
          alt=""
          class="leaflet-marker-shadow leaflet-zoom-hide"
          src="marker-shadow.png"
          style="margin-left: -12px; margin-top: -41px; width: 41px; height: 41px; left: 138px; top: 0px;"
        />
      </div>
      <div
        class="leaflet-pane leaflet-marker-pane"
      >
        <img
          alt="Marker"
          class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
          role="button"
          src="marker-icon.png"
          style="margin-left: -12px; margin-top: -41px; width: 25px; height: 41px; left: -22px; top: 114px; z-index: 114;"
          tabindex="0"
        />
        <img
          alt="Marker"
          class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
          role="button"
          src="marker-icon.png"
          style="margin-left: -12px; margin-top: -41px; width: 25px; height: 41px; left: 138px; top: 0px; z-index: 0;"
          tabindex="0"
        />
      </div>
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-disabled="false"
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            <span
              aria-hidden="true"
            >
              +
            </span>
          </a>
          <a
            aria-disabled="false"
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            <span
              aria-hidden="true"
            >
              −
            </span>
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JavaScript library for interactive maps"
          >
            Leaflet
          </a>
           
          <span
            aria-hidden="true"
          >
            |
          </span>
           © OpenStreetMap
        </div>
      </div>
    </div>
  </div>
</div>
`;
